### DISCUSSION TITLE
rclcpp 예제 작성 전 상의하고 싶은 부분

### DISCUSSION THREAD

**Discussion by routiful** [robotpilot/ros2-seminar-examples#7] [Ideas]
@robotpilot 
rclcpp 예제 작성 전에 몇가지 상의하고 싶은 부분이 있습니다.

### rclpy example 폴더 구조

└── topic_service_action_rclpy_example
    ├── arithmetic
        ├── argument.py
        ├── __init__.py
        ├── operator.py
    ├── calculator
        ├── calculator.py
        ├── __init__.py
        ├── main.py
    ├── checker
        ├── checker.py
        ├── __init__.py
        └── main.py

## main 문의 위치
calculator 와 checker의 경우 `main.py`가 있는데 arithmetic에는 없네요. arithmetic에도 main 파일을 따로 만드는게 통일성 있지 않을까요?

## service client 예제
service client 를 timer를 이용해서 주기적으로 request, response 하는 예제는 service 통신의 개념과 그 목적에 맞지 않는 것 같습니다.
operator.py에서 timer는 삭제하고 해당 노드 실행 시 한번의 request, response를 실행하도록 하는건 어떨까요?

## action server, client 예제 네이밍 변경
사칙연산 결괏값을 모두 더하고 request 값과 비교하는 것이 해당 메시지 통신의 목적이니 그 이름을 `arithmetic_checker` 가 아닌  `sum_integrator` 로 변경하는건 어떨까요?

## 예제 코드 안의 주석
코드 안에 주석이 필요한 부분이 생길 수 있을 것 같은데, 그때는 영어로 주석을 달아도 괜찮을까요? 아니면 어차피 글로 설명할 예정이니 주석은 따로 달지 않는걸로 할까요?

## rclcpp 코드와 rclpy 코드간의 상호호환성
이미 작성하신 rclpy 코드의 클래스, 함수, 변수 등이 똑같은 rclcpp 예제를 작성하여 두 예제가 쉽게 읽힐 수 있도록 하는게 최종적으로는 좋을 것 같긴한데, 메시지 통신을 위한 인터페이스는 유지하면서 내부 코드 디자인은 조금 변경해도 괜찮을까요?
폴더 구조는 rclpy 예제와 똑같이 하면서 내부 코드는 제가 자유롭게 개발하고 추후 리뷰를 통해 rclpy 예제와 다시 맞춰가고 싶어서 제안 드립니다.

감사합니당! :)

**Comment by robotpilot** [comment]
# main 문의 위치
- 일부로 다르게 구성해본 것입니다. 다르게 해야 설명할 수 있는 부분 setup.py, launch 등이 있어서요.
- 이대로 가기를 희망합니다.

# service client 예제
- 그러네요. 변경합시다.

# action server, client 예제 네이밍 변경
- 변경 자체는 좋습니다만 `sum_integrator` 은 맞나요? 전자공학 배우던 시절이 떠오르네요. 더 적절한 이름이 있다면 바꾸겠습니다.

# 예제 코드 안의 주석
- 예제 코드 안의 주석을 달아본 적이 있는데 주석은 항상 바꾸고 싶어지더라고요. 그러면 나중에 책으로 엮을때 힘들어집니다. 
- 가능한 주석은 없애고 강좌와 책에서 서술하면 좋을 듯 싶네요.
 
# rclcpp 코드와 rclpy 코드간의 상호호환성
- 네. 가능한 호환성을 가지면 좋죠. 제 쪽도 변경이 필요하면 적극적으로 변경하겠습니다.

**Reply by routiful** [reply]
답변 감사합니다! 이번 주말내로 PR 하고 글을 올려보도록 하겠습니다 :)

**Reply by robotpilot** [reply]
@routiful https://github.com/robotpilot/ros2-seminar-examples/pull/8 에서 일부 반영 하였습니다.

**Comment by bmaxdk** [comment]
ROS2 에서  shared pointer 를 사용 하는데 main class 에서 node 를 만듭니다. 근데 다른 class 를 만들어서 거기서 같은 node를 사용 할수 있는데 두개의 다른 클래스에서 같은 node를 share 할때 pros and cons 또는 언제 주로 사용하는지와 structure 을 detail하게 설명 해주실수 있을까요?
